src/mcts/search.cc

```diff
@@ -1507,9 +1507,26 @@ void SearchWorker::DoBackupUpdateSingleNode(
       d = n->GetD();
       m = n->GetM();
     }
+    const auto before_visits = n->GetN();
     n->FinalizeScoreUpdate(v / (1.0f + params_.GetShortSightedness() * depth),
                            d, m, node_to_process.multivisit);
 
+    // Flatten policy if we've visited the node enough times. We efficiently
+    // sample every 2^13 visits as an approximation to 10k visits. Because
+    // multivisits is significantly smaller than the sampling rate, we can check
+    // when the appropriate bit changes.
+    const auto after_visits = n->GetN();
+    const auto ratio = params_.GetPolicyFlattenRatio();
+    if (ratio && (before_visits ^ after_visits) & 1 << 13) {
```

Comment:
@ddobbelaere suggested checking before and after visits to handle multivists. Here's using xor to check if it flips. Optimizer has put the float ratio check first now.
```asm
        movl    64(%rbx), %r15d
…
        callq   __ZN6lczero4Node19FinalizeScoreUpdateEfffi
…
        movl    64(%rbx), %ecx
        movq    96(%rdi), %rax
        vmovss  64(%rax), %xmm0         ## xmm0 = mem[0],zero,zero,zero
        vucomiss        LCPI56_2(%rip), %xmm0
        jne     LBB56_12
        jnp     LBB56_22
LBB56_12:                               ##   in Loop: Header=BB56_11 Depth=1
        xorl    %ecx, %r15d
        testl   $8192, %r15d            ## imm = 0x2000
        je      LBB56_22
```

Reply:
Actually… I believe `node_to_process.multivisit` is always 1 in most of `DoBackupUpdateSingleNode` so we can revert the need to handle multivisit > 1. I'll revert and update the comment.

Reply:
We removed multivisits to terminals because of incorrect calculation of how many visits was reasonable - but it is possible we may bring it back at some future point.
Note that the xor flip checking doesn't always work if you multivisit 8192 or more :P

Reply:
Right. To be clear, the current behavior of the main body of `DoBackupUpdateSingleNode` is to always have multivisit = 1 because the > 1 multivisit only exist as Collision right now and those are handled separately. If we do end up with multivisit > 1 in the body, we can restore the XOR where I had in the comment "Because multivisits is significantly smaller than the sampling rate, we can check when the appropriate bit changes."

But if we allow multivisits to approach/exceed the sampling rate, we have various options:
- increase the number for sampling rate
- more correctly handle multivisit
- just ignore if most multivisits will be less than the sampling rate anyway.